File: ADT/stack-queue-pblm/132.py
# Meta, Amazon
# Monotonic Stack (Increasing or Decreasing order)

# Tc = O(n)
# Sc = O(n) [stack]
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        n = len(nums)
        num3 = float("-inf")
        st = []
        for i in range(n - 1, -1, -1):
            if nums[i] < num3:
                return True
            while st and st[-1] < nums[i]:
                num3 = st.pop()
            st.append(nums[i])
        return False

File: ADT/stack-queue-pblm/132.js
/**
 * @param {number[]} nums
 * @return {boolean}
 */
var find132pattern = function (nums) {
    const n = nums.length
    let num3 = -Infinity
    const st = []
    for (let i = n - 1; i >= 0; i--) {
        if (nums[i] < num3) {
            return true
        }
        while (st.length > 0 && st[st.length - 1] < nums[i]) {
            num3 = st.pop()
        }
        st.push(nums[i])
    }
    return false
};